server/src/main/java/org/opensearch/extensions/action/ExtensionActions.java

```diff
@@ -0,0 +1,149 @@
+import org.opensearch.extensions.ExtensionsOrchestrator;
+import org.opensearch.extensions.RegisterTransportActionsRequest;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponse;
+import org.opensearch.transport.TransportResponseHandler;
+import org.opensearch.transport.TransportService;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * This class manages TransportActions for extensions
+ *
+ * @opensearch.internal
+ */
+public class ExtensionActions {
+    private static final Logger logger = LogManager.getLogger(ExtensionActions.class);
+    private Map<String, DiscoveryExtension> actionsMap;
+    private final Map<String, DiscoveryExtension> extensionIdMap;
+    private final TransportService transportService;
+    private final NodeClient client;
+
+    public ExtensionActions(Map<String, DiscoveryExtension> extensionIdMap, TransportService transportService, NodeClient client) {
+        this.actionsMap = new HashMap<>();
+        this.extensionIdMap = extensionIdMap;
+        this.transportService = transportService;
+        this.client = client;
+    }
+
+    void registerAction(String action, DiscoveryExtension extension) throws IllegalArgumentException {
+        if (actionsMap.containsKey(action)) {
+            throw new IllegalArgumentException("The " + action + " you are trying to register is already registered");
+        }
+        actionsMap.putIfAbsent(action, extension);
+    }
+
+    public DiscoveryExtension getExtension(String action) {
+        return actionsMap.get(action);
+    }
+
+    /**
+     * Handles a {@link RegisterTransportActionsRequest}.
+     *
+     * @param transportActionsRequest  The request to handle.
+     * @return  A {@link ExtensionBooleanResponse} indicating success.
+     */
+    public TransportResponse handleRegisterTransportActionsRequest(RegisterTransportActionsRequest transportActionsRequest) {
+        /*
+         * We are proxying the transport Actions through ExtensionMainAction, so we really dont need to register dynamic actions for now.
+         */
+        logger.debug("Register Transport Actions request recieved {}", transportActionsRequest);
+        DiscoveryExtension extension = extensionIdMap.get(transportActionsRequest.getUniqueId());
+        for (String action : transportActionsRequest.getTransportActions().keySet()) {
+            registerAction(action, extension);
+        }
+        return new ExtensionBooleanResponse(true);
+    }
+
+    public TransportResponse handleTransportActionRequestFromExtension(TransportActionRequestFromExtension request) {
+        DiscoveryExtension extension = extensionIdMap.get(request.getUniqueId());
+        final CountDownLatch inProgressLatch = new CountDownLatch(1);
```

Comment:
I would expect this to be paired with an `await()` call somewhere (with a timeout in case the internet died).  I don't see this and thus don't see the purpose for this latch.

Reply:
Sure probably a miss, thanks for the call out. Taken care of it.